main/mbcs.c

```diff
@@ -0,0 +1,115 @@
+/*
+*   $Id$
+*
+*   Copyright (c) 1996-2003, Darren Hiebert
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License.
+*
+*   This module contains functions for checking multibyte character set.
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#define __USE_GNU
+#include "general.h"  /* must always come first */
+
+#ifdef HAVE_ICONV
+
+#include <stdio.h>
+#include <string.h>
+#include <iconv.h>
+#include <errno.h>
+#include "options.h"
+#include "mbcs.h"
+#include "routines.h"
+
+static iconv_t iconv_fd = (iconv_t) -1;
+
+extern boolean openConverter (char* encoding)
+{
+	if (!encoding)
+		return FALSE;
+	iconv_fd = iconv_open("UTF-8", encoding);
+	return iconv_fd != (iconv_t) -1;
+}
+
+extern boolean isConverting ()
+{
+	return iconv_fd != (iconv_t) -1;
+}
+
+extern boolean convertString (vString *const string)
+{
+	size_t utf8_len = 0, mbcs_len, utf8_size;
+	char *utf8, *mbcs, *utf8ptr;
+	if (iconv_fd == (iconv_t) -1)
+		return FALSE;
+	mbcs_len = vStringLength (string);
+	utf8_len = mbcs_len * 4;
+	utf8ptr = utf8 = xCalloc (utf8_len, char);
+	if (!utf8)
+		return FALSE;
+	mbcs = vStringValue (string);
+retry:
+	if (iconv (iconv_fd, &mbcs, &mbcs_len, &utf8ptr, &utf8_len) == (size_t) -1)
+	{
+		eFree (utf8);
+		return FALSE;
+	}
+	if (errno == EILSEQ)
+	{
+		*utf8ptr++ = '?';
+		utf8_len--;
+		mbcs++;
+		mbcs_len--;
+		verbose ("  Encoding: %s\n", strerror(errno));
+		goto retry;
+	}
+
+	utf8_size = utf8ptr - utf8;
+
+	vStringClear (string);
+	while (vStringSize (string) <= utf8_size + 1)
+		vStringAutoResize (string);
+	memcpy (vStringValue (string), utf8, utf8_size + 1);
+	vStringLength (string) = utf8_size;
+	eFree (utf8);
+
+	iconv (iconv_fd, (void*)0, (void*)0, &utf8ptr, &utf8_len);
```

Comment:
hum, what? 

> A  different case is when inbuf is `NULL` or `*inbuf` is `NULL`, but `outbuf` is not `NULL` and `*outbuf` is not `NULL.`  In this case, the `iconv()` function attempts to set `cd`'s conversion  state to  the  initial  state and store a corresponding shift sequence at `*outbuf`.  At most `*outbytesleft` bytes, starting at `*outbuf`, will be written.  If the output buffer  has  no  more   room  for  this reset sequence, it sets `errno` to `E2BIG` and returns `(size_t) -1`.  Otherwise, it increments `*outbuf` and decrements `*outbytesleft` by the number of bytes written.

Don't you rather want this one?: (e.g. all `NULL`s)

> A third case is when `inbuf` is `NULL` or `*inbuf` is `NULL`, and `outbuf`  is  `NULL`  or  `*outbuf`  is  `NULL.`  In this case, the `iconv()` function sets `cd`'s conversion state to the initial state.